### PR TITLE
KAFKA-10020: Create a new version of a scala Serdes without name clash (KIP-616)

### PR DESCRIPTION
**Pull request by LMnet**
Implementation of a solution for KIP-616.

Wildcard import of the old `org.apache.kafka.streams.scala.Serdes` leads
to a name clash because some of implicits has the same names as types
from the scala's std lib. New
`org.apache.kafka.streams.scala.serialization.Serdes` is the same as old
`Serdes`, but without name clashes. The old one was marked as deprecated.

Also, missing serdes for `UUID`, `ByteBuffer` and `Short` types are present in
the new `Serdes`.

### CODE REVIEW

**Comment by LMnet** [comment]
@vvcephei @mjsax

**Comment by LMnet** [comment]
I added `nullSerde` and rebased branch to the current trunk.

**Comment by LMnet** [comment]
Voting for the KIP successfully finished and this pull request could be merged now.

**Comment by mjsax** [comment]
Retest this please.

**Comment by LMnet** [comment]
There were some failed style checks. I fixed them.

**Comment by bbejeck** [comment]
Ok to test

**Comment by mjsax** [comment]
Retest this please.

**Comment by mjsax** [comment]
Retest this please.

**Comment by mjsax** [comment]
Retest this please.

**Comment by mjsax** [comment]
Arg... Jenkins seems to be on holidays... Will try again later.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/Serdes.scala
Comment (vvcephei) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/Serdes.scala]: Can you add the "since" parameter?

**Review by vvcephei — COMMENTED** [review]
Thanks for the PR, @LMnet !

Should we have some test coverage for the new class?

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/Serdes.scala
Comment (LMnet) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/Serdes.scala]: Done

**Comment by LMnet** [comment]
@vvcephei I completely forgot about tests, thanks for reminding me! I changed all usages of the old `org.apache.kafka.streams.scala.Serdes` to the new `org.apache.kafka.streams.scala.serialization.Serdes` in tests. I think it should be enough in terms of test coverage.

Also, I fixed `KGroupedStream`, `SessionWindowedKStream`, `TimeWindowedKStream`, and docs: they used the old `Serdes` object.

I was not sure about commits policy in Kafka so I added all these changes in the separate commit. I can squash it later if needed.

**Comment by vvcephei** [comment]
Thanks, @LMnet ! 

The commit policy is just that the committer will squash the whole PR into one commit when merging, so you don't have to worry about it.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/serialization/Serdes.scala
Comment (vvcephei) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/serialization/Serdes.scala]: Ooh, this one might become a problem.

See https://cwiki.apache.org/confluence/display/KAFKA/KIP-659%3A+Improve+TimeWindowedDeserializer+and+TimeWindowedSerde+to+handle+window+size

The fact that we're deprecating this constructor isn't what concerns me, it's the reason why: if you were to implicitly use this serde to _read_ records, you'd get incorrect results. This constructor fills in the end of the window as `Long.MAX_VALUE`, which will be wrong approximately all the time.

Further, when we pass serdes like this, they have to be pre-configured. Streams won't call `configure`. We could plan to require an `implicit streamsConfig: Properties` argument and then extend KIP-659 to configure this serde using the new `window.size.ms` config. But that might not be the right window size in practice either, and the fact that it's an implicit makes me concerned that it would be really hard to track down the bug in user code.

I'm sort of thinking maybe we should just drop this one implicit, since you really need to specify the window size for it to work properly.

Note, the _serializer_ is safe, it's only the _deserializer_ that we need to worry about. So, my other idea would be for the implicit to pack up a new serde with the `new TimeWindowedSerializer(inner)` and some kind of `new ThrowUnsupportedOperationExceptionDeserializer()` that prevents the implicit from being used inappropriately. The exception message could say "This implicit serde is only for serializing keys. To deserialize as well, you should directly provide a WindowedSerdes.TimeWindowedSerde instance."

What do you think?

**Review by vvcephei — COMMENTED** [review]
Thanks @LMnet !

This looks good to me. Unfortunately, I just now realized one thing about TimeWindowedSerde (precipitated by the KIP-659 discussion). We can discuss what you want to do here, and then amend the KIP before merging.

Thanks again,
-John

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/serialization/Serdes.scala
Comment (LMnet) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/serialization/Serdes.scala]: I got acquainted KIP-659 and in my opinion, there are few ways to handle this problem:
1. Delete `timeWindowedSerde` from the new `Serdes` object, because it's a new code and we would not have any compatibility issues. Also, we could add some comments in the deprecation message about this serde.
2. Add another implicit requirement to the `timeWindowedSerde`. For example `implicit streamsConfig: Properties`, like you suggested. But `Properties` is too wide and not specific type. I personally don't like this option. I think it could become one more kafka pitfall.
3. Remove implicit modifier from `timeWindowedSerde`. We could make it a regular function that requires streams config explicitly.

I prefer the third option.

A bit of offtop, but I found all idea of configurable through config serdes confusing. I have to describe serdes in one place, but config them in some other place. I would prefer a system that has immutable and not configurable serdes with an extra layer on top of it with configurations. Serdes mutability already makes some difficulties and that's why in `Serdes` object all instances are `def`, not `val`.

Comment (vvcephei) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/serialization/Serdes.scala]: Thanks @LMnet ,

I agree with you about option 2, it seems unlikely to actually work automatically and very likely to result in confusing implicit resolution conflict compiler errors.

(1) and (3) are both fine with me. (3) seems mildly unnecessary, since it would just pass through directly to the constructor of the TimeWindowedSerde. Just to be clear, though, I would _not_ recommend :

```scala
def timeWindowedSerde[T](tSerde: Serde[T], properties: Properties): WindowedSerdes.TimeWindowedSerde[T]
```

Rather, I'd suggest:
```scala
def timeWindowedSerde[T](tSerde: Serde[T], windowSize: Long): WindowedSerdes.TimeWindowedSerde[T] = new WindowedSerdes.TimeWindowedSerde[T](tSerde, windowSize)
```

I.e., if you're explicitly constructing the serde anyway, you might as well explicitly specify its parameters. The config properties are really for the case when the serde is reflectively constructed by the Consumer.

By the way, I agree about the idea of configuring serdes to be confusing. It's really to support more complex serdes that store their schemas externally AND that are constructed reflectively by Streams. For example, instead of passing the serdes into the DSL at all (which is what these implicits automate for Scala users), you might just set `default.value.serde` to a "json serde" or an "avro serde" or a "schema registry serde". These things need access to the runtime config, and the mechanism to give it to them is to reflectively construct them via the zero-arg constructor and then invoke `configure()` with the runtime configs. On the other hand, the serdes in _this_ file, like UUIDSerde or LongSerde, don't need any configuration; they already fully specify their schemas with no additional input. The TimeWindowedSerde does need some additional input (the inner type and the window size). When it's constructed reflectively, this information has to be passed to the serde via configs, but when it's constructed by user code, we can (and should) just pass all the information to the constructor.

Comment (LMnet) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/serialization/Serdes.scala]: Ok, I agree that the third option is a bit unnecessary. We already have a constructor for that.

I deleted `WindowedSerdes` at all and mentioned them in the deprecated message.

**Review by vvcephei — APPROVED** [review]
Thanks @LMnet !

Since PR builds are broken, I'll test this and merge it.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/serialization/Serdes.scala
Comment (vvcephei) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/serialization/Serdes.scala]: Ah, we don't technically need to drop this one, only the `timeWindowedSerde`. Sorry for the confusion.

Comment (vvcephei) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/serialization/Serdes.scala]: Adding it back will also unblock the build, which fails right now because `WindowedSerdes` is an unused import. I'll just add this back and push to your branch, assuming the rest of the tests pass, so I can go ahead and merge it.

**Comment by vvcephei** [comment]
Thanks so much for the contribution and your patience with our long discussion, @LMnet !

**Comment by chia7712** [comment]
```
[Warn] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala:22: imported `Serdes' is permanently hidden by definition of object Serdes in package scala
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala:136: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala:137: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala:142: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala:142: value javaLongSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala:145: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala:145: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala:153: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala:153: value javaLongSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala:153: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala:163: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala:163: value javaLongSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala:167: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestImplicitSerdes.scala:167: value javaLongSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Warn] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:45: imported `Serdes' is permanently hidden by definition of object Serdes in package scala
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:358: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:358: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:368: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:368: value intSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:370: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:370: value intSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:377: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:377: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:379: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:379: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:384: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:384: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:384: value intSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:391: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:391: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:391: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:414: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:414: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:424: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:426: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:433: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:433: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:442: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:442: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:450: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:450: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Warn] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala:401: local val keyValueMapper in method getTopologyJava is never used
[Warn] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/WordCountTest.scala:28: imported `Serdes' is permanently hidden by definition of object Serdes in package scala
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/WordCountTest.scala:147: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/WordCountTest.scala:148: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/WordCountTest.scala:165: value stringSerde is not a member of object org.apache.kafka.streams.scala.Serdes
[Error] /home/chia7712/kafka/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/WordCountTest.scala:165: value javaLongSerde is not a member of object org.apache.kafka.streams.scala.Serdes
four warnings found
43 errors found
```

It seems this PR breaks the build with JDK8 + SCALA-2.12. @LMnet Could you please take a look?

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/serialization/Serdes.scala
Comment (LMnet) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/serialization/Serdes.scala]: Ok, thanks!

**Comment by LMnet** [comment]
@chia7712 ok, I will check this. I think I know how to fix this.

**Comment by LMnet** [comment]
@chia7712 I created a fix: https://github.com/apache/kafka/pull/9245

**Comment by vvcephei** [comment]
Thanks for pointing that out, @chia7712 , and sorry for breaking the build.

**Pull request merged**
